bin/flutter-dev

```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env bash
+# Copyright 2014 The Flutter Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+set -e
+
+# This is a helper script for development purposes. It runs the Flutter tool
+# from source code directly, without using the prebuilt snapshot. This is
+# useful for development, as it allows you to make changes to the tool and see
+# the effects immediately, but is much slower than using the prebuilt snapshot.
+
+# To debug the tool, you can uncomment the following lines to enable debug
+# mode and set an observatory port:
+# FLUTTER_TOOL_ARGS="--enable-asserts $FLUTTER_TOOL_ARGS"
+# FLUTTER_TOOL_ARGS="$FLUTTER_TOOL_ARGS --observe=65432"
+
+# Needed because if it is set, cd may print the path it changed to.
+unset CDPATH
+
+# On Mac OS, readlink -f doesn't work, so follow_links traverses the path one
+# link at a time, and then cds into the link destination and find out where it
+# ends up.
+#
+# The returned filesystem path must be a format usable by Dart's URI parser,
+# since the Dart command line tool treats its argument as a file URI, not a
+# filename. For instance, multiple consecutive slashes should be reduced to a
+# single slash, since double-slashes indicate a URI "authority", and these are
+# supposed to be filenames. There is an edge case where this will return
+# multiple slashes: when the input resolves to the root directory. However, if
+# that were the case, we wouldn't be running this shell, so we don't do anything
+# about it.
+#
+# The function is enclosed in a subshell to avoid changing the working directory
+# of the caller.
+function follow_links() (
+  cd -P "$(dirname -- "$1")"
+  file="$PWD/$(basename -- "$1")"
+  while [[ -h "$file" ]]; do
+    cd -P "$(dirname -- "$file")"
+    file="$(readlink -- "$file")"
+    cd -P "$(dirname -- "$file")"
+    file="$PWD/$(basename -- "$file")"
+  done
+  echo "$file"
+)
+
+# Lookup the parent directory of the script (../), the Flutter SDK root.
+PROG_NAME="$(follow_links "${BASH_SOURCE[0]}")"
+BIN_DIR="$(cd "${PROG_NAME%/*}" ; pwd -P)"
+FLUTTER_ROOT="$(cd "$BIN_DIR/.." ; pwd -P)"
+
+# Run the Flutter tool directly without using the cached snapshot.
+exec "$FLUTTER_ROOT/bin/cache/dart-sdk/bin/dart" \
```

Comment:
Hmm, this will not cache the dart sdk if it's not already present (or more importantly, if the cached dart sdk is a different version from what the current engine.version is).

Instead, I would have this file source //flutter/bin/internal/shared.sh like `//flutter/bin/dart` and `//flutter/bin/flutter` do: https://github.com/flutter/flutter/blob/master/bin/dart#L56-L59

And then at the end of shared.sh, when we switch over the `$BIN_NAME`, add a case for `flutter-dev`: https://github.com/flutter/flutter/blob/master/bin/internal/shared.sh#L256-L269

I would also add `flutter-dev` to the early exit in the `upgrade_flutter()` function so that we don't compile the snapshot (since we're running from source anyway): https://github.com/flutter/flutter/blob/master/bin/internal/shared.sh#L146-L149

Reply:
Done!